source/fuzz/transformation_add_dead_break.cpp

```diff
@@ -0,0 +1,289 @@
+                                         bb_from->terminator())) {
+        // The given id is no good as its definition does not dominate the exit
+        // of |from_block_|
+        return false;
+      }
+    }
+    phi_index++;
+  }
+  // Reject the transformation if not all of the ids for extending OpPhi
+  // instructions are needed. This might turn out to be stricter than necessary;
+  // perhaps it would be OK just to not use the ids in this case.
+  return phi_index == phi_ids_.size();
+}
+
+bool TransformationAddDeadBreak::FromBlockIsInLoopContinueConstruct(
+    IRContext* context, uint32_t maybe_loop_header) {
+  // We deem a block to be part of a loop's continue construct if the loop's
+  // continue target dominates the block.
+  auto containing_construct_block = context->cfg()->block(maybe_loop_header);
+  if (containing_construct_block->IsLoopHeader()) {
+    auto continue_target = containing_construct_block->ContinueBlockId();
+    if (context->GetDominatorAnalysis(containing_construct_block->GetParent())
+            ->Dominates(continue_target, from_block_)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool TransformationAddDeadBreak::AddingBreakRespectsStructuredControlFlow(
+    IRContext* context, BasicBlock* bb_from) {
+  // Look at the structured control flow associated with |from_block_| and
+  // check whether it is contained in an appropriate construct with merge id
+  // |to_block_| such that a break from |from_block_| to |to_block_| is legal.
+
+  // There are three legal cases to consider:
+  // (1) |from_block_| is a loop header and |to_block_| is its merge
+  // (2) |from_block_| is a non-header node of a construct, and |to_block_|
+  //     is the merge for that construct
+  // (3) |from_block_| is a non-header node of a selection construct, and
+  //     |to_block_| is the merge for the innermost loop containing
+  //     |from_block_|
+  //
+  // The reason we need to distinguish between cases (1) and (2) is that the
+  // structured CFG analysis does not deem a header to be part of the construct
+  // that it heads.
+
+  // Consider case (1)
+  if (bb_from->IsLoopHeader()) {
+    // Case (1) holds if |to_block_| is the merge block for the loop;
+    // otherwise no case holds
+    return bb_from->MergeBlockId() == to_block_;
+  }
+
+  // Both cases (2) and (3) require that |from_block_| is inside some
+  // structured control flow construct.
+
+  auto containing_construct =
+      context->GetStructuredCFGAnalysis()->ContainingConstruct(from_block_);
+  if (!containing_construct) {
+    // |from_block_| is not in a construct from which we can break.
+    return false;
+  }
+
+  // Consider case (2)
+  if (to_block_ ==
+      context->cfg()->block(containing_construct)->MergeBlockId()) {
+    // This looks like an instance of case (2).
+    // However, the structured CFG analysis regards the continue construct of a
+    // loop as part of the loop, but it is not legal to jump from a loop's
```

Comment:
It is legal, but only from the back-edge block. I'm not sure if this is a worthwhile distinction here.

Reply:
It would be good to allow this case eventually; I have filed #2577 to log it, and added a TODO(2577) comment.